File: 백준/Gold/2629. 양팔저울/양팔저울.py
import sys
input = sys.stdin.readline

def sol(num, wg):
    if num > weight_num: # 구슬의 숫자가 더 큰 경우
        return
    if dp[num][wg] == 1: # 이미 연산함
        return
    dp[num][wg] = 1
    
    sol(num + 1, wg + weight[num-1])       # 추 추가하기
    sol(num + 1, wg)                       # 그대로 진행
    sol(num + 1, abs(wg - weight[num-1]))  # 반대편에 추 추가(무게빼기)

weight_num = int(input())
weight = list(map(int, input().split()))
bead_num = int(input())
bead = list(map(int, input().split()))
# dp[사용한 추의 개수][무게 차이]
dp = [[0 for _ in range(500 * weight_num + 1)] for _ in range(weight_num + 1)]

sol(0, 0)

for k in bead:
    if k > 500 * weight_num: # 가능한 구슬 무게 범위 초과
        print("N", end=' ')
    elif dp[weight_num][k]:  # 구슬 무게 측정 가능
        print("Y", end=' ')
    else:                    # 구슬 무게 측정 불가능
        print("N", end=' ')

File: 백준/Gold/2629. 양팔저울/양팔저울.cc
#include <iostream>
#include <cmath>
using namespace std;
int weight_num, bead_num;
int weight[31] = {0, };
int bead[8] = {0, };
// dp[사용한 추의 개수][무게 차이]
int dp[31][15001] = {0, };

void sol(int num, int wg) {
    if(num > weight_num) // 구슬의 숫자가 더 큰 경우
        return;
    if(dp[num][wg] == 1) // 이미 연산함
        return;
    dp[num][wg] = 1;

    sol(num + 1, wg + weight[num]);         // 추 추가하기
    sol(num + 1, wg);                       // 그대로 진행
    sol(num + 1, abs(wg - weight[num]));    // 반대편에 추 추가(무게빼기)
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    
    cin >> weight_num;
    for(int i = 0; i < weight_num; i++)
        cin >> weight[i];

    cin >> bead_num;
    for(int i = 0; i < bead_num; i++)
        cin >> bead[i];

    sol(0, 0);

    for(int i = 0; i < bead_num; i++) {
        if(bead[i] > 500 * weight_num)
            cout << "N" << " ";
        else if(dp[weight_num][bead[i]] == 1)
            cout << "Y" << " ";
        else
            cout << "N" << " ";
    }

    return 0;
}